### PR TITLE
Add BindCraft-style optimizer

### PR DESCRIPTION
**Pull request by magnusross**
This PR adds the BindCraft optimisation pipeline. You can see an overview of their methodology in [the paper](https://www.biorxiv.org/content/10.1101/2024.09.30.615802v1.full.pdf). BindCraft itself is based on colabdesign so lots of the steps are actually ported from there. 

We add two public functions:
- `colabdesign_stage` implements a single stage of colabdesign optimisation. The bindcraft optimisation is made up of 4 such stages. Each stage ramps/anneals the temperature/learning rate between two specified values. In order to jit without recompling the loss each time these change i put the loss in an equinox module. The loss is a composite of different terms depending on if `hard` is true. If so we use the straight thru estimator on the argmaxed sequence, if not we use the pssm mixed with the raw logits, with the proportion pssm being given by `soft` param. Note that when soft is less than 1.0, we will be feeding unconstrained/negative values into the model. The step also does some gradient normalisation.
- `bindcraft_design` this function mirrors the bindcraft optimisation pipeline from their defaults. There are 4 separate stages of continous optimisation. We also copy them in using  gaussian distributed initialisation for the logits. We do not implement the semigreedy discrete optimisation stage. 

Some thing I am unsure about:
- Perhaps, to avoid adding too much code, it might be better to not include the bindcraft_design fuction, and let the user implement that for themself? Or we could put it as an example in the readme?
- Do we need to worry that some models might error when the logits are fed in?
- Should I implement some tests?

### CODE REVIEW

**Comment by nboyd** [comment]
hey @magnusross thanks, this is very nice! 

Could you implement the pullback for _ColabDesignLoss separately? You could do this manually like I do in [`biohub_optimizer`](https://github.com/escalante-bio/mosaic/blob/46301039d7019b70d0754456420b449a1da6b366/src/mosaic/optimizers.py#L671), or, perhaps, a much better option would be to use jax.vjp or grad. The basic idea here is to make sure the user can switch optimizers without recompilation (so we only call `_eval_loss_and_grad` on the user's loss, rather than the precomposition with _ColabDesignLoss). 

`bindcraft_design` looks fine to me but could you make `x` an argument rather than hardcoding the initialization? it might also be nicer to have stages be a list of dictionaries that you then splat into colabdesign_stage (or if you want to be fancy it could be a list of functools.partial). It's a bit more verbose but hopefully makes it more clear what those arguments are.

**Comment by magnusross** [comment]
Thanks for the review @nboyd :)

I have implemented both suggestions, vjp thing now makes sense, it's taking a bit of time to get my head around jax compilation! 

Let me know if there is anything else.

**Comment by nboyd** [comment]
Looks great, thanks! Rebase on main and we'll merge 😀
